### PR TITLE
Refactor the way categories are attached to Addons using StaticCategory

### PR DESCRIPTION
**Pull request by diox**
This makes `all_categories`, `app_categories`, `get_category` coherent, using the same source, so that it's easier to then override when needed. `app_categories` is changed to return a dict, because it's
more useful that way.

This removes the need for the special `attach_categories()` function that the add-on indexer was using, as we now load all categories for each add-on instead of just the first one.

The final (big!) change is to use `StaticCategory` instances instead of `Category` for these properties, because we have all the relevant information in the constants.

### CODE REVIEW

File: src/olympia/addons/forms.py [hunk 1]
```diff
@@ -208,21 +208,19 @@ class CategoryForm(forms.Form):
...
 
     def save(self, addon):
-        application = self.cleaned_data.get('application')
-        categories_new = self.cleaned_data['categories']
-        categories_old = [cats for app, cats in addon.app_categories if
-                          (app and application and app.id == application) or
-                          (not app and not application)]
-        if categories_old:
-            categories_old = categories_old[0]
+        application = self.cleaned_data.get('application', amo.FIREFOX.id)
```
Comment (eviljeff) [inline src/olympia/addons/forms.py]: hmm.. where are we saving the form without `application` being sent?

File: src/olympia/addons/tests/test_models.py [hunk 2]
```diff
@@ -1102,60 +1102,72 @@ def test_newlines_less_than_tight(self):
...
 
-        c = Category(application=amo.THUNDERBIRD.id, db_name='XXX',
-                     type=addon().type, count=1, weight=1)
-        c.save()
-        AddonCategory.objects.create(addon=addon(), category=c)
-        c24.save()  # Clear the app_categories cache.
-        app_cats += [(amo.THUNDERBIRD, [c])]
-        assert addon().app_categories == app_cats
+        thunderbird_static_cat = (
+            CATEGORIES[amo.THUNDERBIRD.id][amo.ADDON_EXTENSION]['tags'])
```
Comment (eviljeff) [inline src/olympia/addons/tests/test_models.py]: This test is hard to follow with the setup for Thunderbird in a different style than Firefox and half-way through the method.  Can you move it further up and make it similar?

File: src/olympia/addons/tests/test_models.py [hunk 3]
```diff
@@ -1102,60 +1102,72 @@ def test_newlines_less_than_tight(self):
...
+
+        # Associate this add-on with a couple more categories, including
+        # one that does not exist in the constants.
+        c2 = Category.objects.create(application=amo.SUNBIRD.id, id=123456,
                                      type=amo.ADDON_EXTENSION, name='Sunny D')
         AddonCategory.objects.create(addon=addon, category=c2)
-
-        # Sunbird category should be excluded.
-        assert get_addon().app_categories == [(amo.FIREFOX, cats)]
+        static_thunderbird_category = (
```
Comment (eviljeff) [inline src/olympia/addons/tests/test_models.py]: same as test above
 (if there were more than 2 tests I'd say make it a separate class and move the setup to `setUp` but meh)

**Comment by eviljeff** [comment]
r+wc

File: src/olympia/addons/forms.py (hunk 1 above)
Comment (diox) [inline src/olympia/addons/forms.py]: Hopefully nowhere, but it's possible since it's not `required`. The old fallback was `None`, which is not great. Re-reading this though, I should replace `addon.app_categories[]` below by add a `addon.app_categories.get()` instead - the goal here is simply to avoid failing catastrophically if `application` is absent.

File: src/olympia/addons/tests/test_models.py (hunk 2 above)
Comment (diox) [inline src/olympia/addons/tests/test_models.py]: Yeah, I'll do that.
